tensorboard/webapp/widgets/line_chart_v2/lib/renderer/svg_renderer.ts

```diff
@@ -13,17 +13,53 @@ See the License for the specific language governing permissions and
 limitations under the License.
 ==============================================================================*/
 
-import {Polyline, Rect} from '../internal_types';
+import {Point, Polyline, Rect} from '../internal_types';
 import {arePolylinesEqual} from '../utils';
-import {LinePaintOption, ObjectRenderer} from './renderer_types';
-
-interface LineCacheValue {
-  type: 'line';
+import {
+  CirclePaintOption,
+  LinePaintOption,
+  ObjectRenderer,
+  TrianglePaintOption,
+} from './renderer_types';
+
+interface PathCacheValue {
   data: Polyline;
   dom: SVGPathElement;
 }
 
-type CacheValue = LineCacheValue;
+interface CircleCacheValue {
+  dom: SVGCircleElement;
+  data: Point;
+}
+
+type CacheValue = PathCacheValue | CircleCacheValue;
+
+function createOrUpdateObject<T extends SVGPathElement | SVGCircleElement>(
+  prevDom: T | undefined,
+  creator: () => T,
+  updater: (el: T) => T,
+  paintOpt: {visible: boolean; color: string; opacity?: number}
```

Comment:
I want to keep the PaintOption private and since this method only make use of visibility/opacity/color, I want to be explicit of what type we expect here.